### PR TITLE
WIP: Add support for move and resize disk on clones

### PR DESCRIPTION
**Pull request by luhahn**
Added support to resize and move disks to clones

added a few helper functions to check state of async calls to proxmox api

### CODE REVIEW

**Comment by luhahn** [comment]
@danitso-dp as requested i created the pull request so we can discuss this more detailed.

I know my pr is far from perfect so i added it as WIP. However i think it would be nice to have the disk modify features.

If you want to I can integrate those changes into your resource. But i first wanted to discuss this with you :)

**Comment by danitso-dp** [comment]
I'll take a look at the code later today and give some feedback.

**Comment by danitso-dp** [comment]
Hi @luhahn 

The day suddenly became rather long so I didn't get to dive into these changes. I'll get back to it soon.

**Comment by danitso-dp** [comment]
I managed to find a bit of time to go through the code and I have the following suggestions:

- Add support for `disk.size` modifications in the current VM resource
- Add support for `disk.datastore_id` modifications in the current VM resource
- Perhaps add a reboot resource with triggers to be used as an action, if this is still relevant with the additions above?

I don't think I missed anything else in the code but do tell me what you think about implementing the changes this way. I can of course help out :)

**Comment by danitso-dp** [comment]
Hi @luhahn 

Check out #7 as an example of how we can move and resize disks using the VM resource. Please note that the code is currently untested.

**Comment by danitso-dp** [comment]
Hi @luhahn 

No rush. We're all busy nowadays :)

I had actually had the async task checks on my todo but I never got around to it as this provider started out as a POC for a project. However, seeing you take that approach was just a clear sign that it was about time to do some refactoring.

I agree that a reboot resource can be useful and while it's not necessarily seen as something you should do with Terraform, I rather enjoy doing as much as possible with Terraform. Adding such a resource is therefore a go from me.

**Comment by luhahn** [comment]
Sorry, the last few days where very full of tasks for me. I'll hopefully get back to this topic till the end of the week

**Comment by luhahn** [comment]
@danitso-dp 
I've updated this branch with your changes in order to make a few adjustments. I added support for clones to move and resize disks and added a new variable in order to target other disks than scsi disks.

However i want to have a look if we can receive the disk name direktly form proxmox, so we can remove my workaround variable and the user dont has to worry about stupid names

**Comment by luhahn** [comment]
A first test on our cluster was successful :)

**Comment by luhahn** [comment]
First Idea of getting the disk name (names) is to use your GetVM function to query the config of the cloned vm. After that we could get all the needed information and determine if move+resize has to run based on the retrieved information.

What to you think @danitso-dp

**Comment by danitso-dp** [comment]
I think the state check for disks need to be worked on to get this feature implemented as the provider is currently operating based on the disk order. This can be a problem with cloned VMs as the disk order could be unknown yet you know you want to modify a specific disk.

One way of doing this would be the disk name approach you've chosen. However, we then also need to rewrite the state check logic as it assumes that disks are stored in a specific order and might show changes when there are in fact no changes to be applied.

The problem comes down to whether to always require a disk name and have the state check look up the disk by name instead of by index, or have this logic apply only to cloned VMs. I'm a bit undecided about which path to choose :)

**Comment by danitso-dp** [comment]
Alright. I managed to squeeze in some time to think :)

Let's proceed with your idea about naming the disks but let's perhaps call it `interface` instead of `name`, and have the argument remain optional as you're doing now.

For existing disks created with previous versions of the provider, we'll simply use `scsiN` as the `interface` value where `N` is the index they have today (0-N). This way we can modify the code to respect the new argument where a state check will then also ensure that the disk is attached to the right interface or trigger an update to correct it.

How does that sound? @luhahn

**Comment by luhahn** [comment]
For clarification:
shall we keep the index (N) in the interface argument? (btw. i agree that interface is the better name for this attribute :)) 
And the default value should be scsi0?

Another question, should we run the move/resize without any further checks? It will fail if the new resize argument is below the current size. And i'm not sure what happens if we try to move if newDatastoreID == oldDatastoreID.

**Comment by luhahn** [comment]
however you suggestion sounds good :)

**Comment by danitso-dp** [comment]
I know the API will complain about shrinking a disk but other than that, nothing bad happens. However, since we know it'll fail, we might as well check for that action and throw an error in the provider instead.

When it comes to the default interface value, the reason why I considered leaving the default value to an empty string and then do scsiN in the code, is to avoid having to do a breaking change. However, being explicit about it will make everything more predictable and if we add a check to see if multiple disks use the same interface, we can prevent that someone accidentally messes up existing VMs when upgrading to v0.4.0.

**Comment by luhahn** [comment]
Currently working on improvements to this PR. Will come back later this day :)

**Comment by luhahn** [comment]
@danitso-dp 
i had some time today and i worked on a few checks regarding the disk interface variable. I'm currently looking at the created vm and getting the disk info, that way i can check if the disk interface exists and if the new size > old size. Let me know what you think about these changes. 

Currently only done on clone vm as an example.

Also tested on our cluster, which works nicely :)

**Comment by luhahn** [comment]
currently working on embedding the new attribute into the current create vm structure

**Comment by luhahn** [comment]
@danitso-dp 

I've got a question: whats the purpose of resourceVirtualEnvironmentVMCreateCustomDisks. Just to import a file for a disk if fileID is present? Or is there more to it?

**Comment by luhahn** [comment]
Current Version is not tested

**Comment by danitso-dp** [comment]
> 
> 
> @danitso-dp
> 
> I've got a question: whats the purpose of resourceVirtualEnvironmentVMCreateCustomDisks. Just to import a file for a disk if fileID is present? Or is there more to it?

Sorry for the late reply. I've been rather busy :)

To answer your question: the function just imports a disk from a file and nothing more. I created it as a separate function to isolate this feature.

I'll take a look at your changes soon and get back to you with some feedback.

**Comment by lotherk** [comment]
> Alright. I managed to squeeze in some time to think :)
> 
> Let's proceed with your idea about naming the disks but let's perhaps call it `interface` instead of `name`, and have the argument remain optional as you're doing now.
> 

Hi,

I'd like to propose "bus" or "device" instead of "interface" as "interface" is mostly used in network context.

Keep up the good work guys :-)

**Comment by luhahn** [comment]
Just noticed, that all VMs are deleted and build again on update, i missed to adjust the Read funktion to store the devices properly for clones

**Comment by luhahn** [comment]
though this is fixed now, there is still work to be done for ide devices. I'm ignoring this currently

**Comment by danitso-dp** [comment]
Hi @luhahn 

Nice work :)

I'll test the code tomorrow against my own cluster. I've just been swamped with work lately so I haven't gotten around to it yet.

**Comment by luhahn** [comment]
@danitso-dp Hey, have you had time to check the changes :)?

**Comment by danitso-dp** [comment]
Hi @luhahn 

Sorry, not yet. I'll be completing a high priority task tomorrow, and I can then resume my work on this provider.

I'll test your changes on Friday and get back to you.

**Comment by luhahn** [comment]
@danitso-dp No worries, take your time. I found a few more things i wanted to fix :) Lets discuss this when you've got more time !

Stay safe and good luck with your task

**Comment by danitso-dp** [comment]
A little later than planned but I now have time for some testing ;)

I'll give it a go and return with some feedback.

**Comment by danitso-dp** [comment]
@luhahn The new features work but there seem to be a state check issue so it keeps wanting to re-create when using the `example` module. I'll take a look at that and get back to you.

**Comment by danitso-dp** [comment]
@luhahn I had some time to look at the code. The issue is caused by existing disk having automatically been attached to scsi0 and onwards. The new code has introduced the new argument for specifying this. However, for existing disks, the interface will be an empty string and the map lookup fails.

Considering the provider is fairly new, we can probably get away with introducing the new argument as a breaking change (required value) and do an upgrade note about assigning scsi0 and onwards to existing disks. This should prevent existing VMs from being destroyed.

**Comment by luhahn** [comment]
@danitso-dp Thanks for your feedback. 
Yeah, i thought something was wrong with the disk assignments. I'll have a look on that tomorrow, maybe I find something which works nice with and without the new argument. However if I don't find a statisfying solution, I would agree, that we're going to indroduce a breaking change.

**Comment by luhahn** [comment]
Sorry I've got a lot of work recentlich and doing a release till the end of the week, so I guess I will start working on the provider next week

**Comment by danitso-dp** [comment]
@luhahn it's the same for me right now but I should have some spare time during the upcoming weekend. I'll check in on your progress to see, if there's something I can contribute.

**Comment by luhahn** [comment]
@danitso-dp 

Allright, i had time to look into the problem. 

We could solve this but it comes with a risk.

What I've currently done to support backwards compatibility is to check if the disk interface is set. If its not set I default to scsi0 (and count up) I've missed this part in resourceVirtualEnvironmentVMReadCustom to update existing VMs. 

If we're following this approach this can prevent the VMs to be recreated. But if one uses a mix of the new and old approach we're going to be in truble. For example using interface scsi1 with the new parameter and an additional disk without the interface parameter. This will lead to a big confusion in the code.

We could check at the begin if either new or old approach is used, or a mix is used in which case we're going to throw an error. Then we would be save.

What do you think?

**Comment by danitso-dp** [comment]
I would actually prefer to introduce a breaking change to avoid too many workarounds, especially since the provider is still fairly new. In case the new argument is marked as required, we should be able to avoid breaking existing VMs by including the upgrade instructions.

**Comment by luhahn** [comment]
Allright, i will work on the changes today and cleanup the code and documentation after that.

**Comment by luhahn** [comment]
Well I think I'm done. Would be nice if you could review (and test) again.

TLDR: 
The Clone part is currently well tested, since it runs currently in our cluster on a jenkins job. And I think it is fine to use :)



However for your notice, we've run in a provider crash about 3 times now. It is not really reproducable and I'm not sure what the cause is. And I'm also not sure if this is caused by the provider or if our setup has network issues from time to time, which causes a resource not to be initialized correctly.

I will investigate further once the error shows up again.

**Comment by danitso-dp** [comment]
I'll test the changes this weekend and get back to you. I'll also try to run it in a loop to see, if I can make it crash.

**Comment by danitso-dp** [comment]
@luhahn I'll get back to this tomorrow as I'm currently adding a standalone Proxmox VE cluster to this repository to improve test automation. I'll get back to the actual testing soon :)

**Comment by danitso-dp** [comment]
@luhahn I didn't have time to fix the following issue but I'll look at it today or Tuesday:

```
Error: Internal validation of the provider failed! This is always a bug
with the provider itself, and not a user issue. Please report
this bug:

1 error occurred:
        * resource proxmox_virtual_environment_vm: interface: One of optional, required, or computed must be set
```

**Comment by luhahn** [comment]
@danitso-dp have you had some more time to look into that issue? I'll be back on this topic next week, since I've got a new feature to complete, and it has to be finished till the end of the week

**Comment by danitso-dp** [comment]
@luhahn I haven't had time to look at it these past two weeks but I'll try to find some time for it tomorrow.

**Comment by luhahn** [comment]
Well found the error here, I messed up the optional: false stuff and missed setting required: true for the disk interface parameter

**Comment by luhahn** [comment]
@danitso-dp Allright, so the fix works. I ran our cluster multiple times without any failure. Seems the crash I had earlier was network problem related on my side. However I only ran the clone part. Maybe you can test the create new vm part as well :)

**Comment by dlandtwing** [comment]
First of all: Thank you both for your effort into making and extending this provider, it has made my life a lot easier!

I have a question regarding the move part of this PR: What is the use case for a clone followed by a disk move? Doesn't the clone API call already allow you to specify the target datastore of the cloned disk?

I'm asking because the way I'm trying to use it right now is resulting in the following error (tested with the current state of the pull request source branch at the time of writing, b6fc2cfec9e65d6bdef8e2ff0d7bb0f96f5850eb):
`Error: Received an HTTP 500 response - Reason: you can't move to the same storage with same format`

I've created a sample tf file with which I can reproduce the error: [main.tf.txt](https://github.com/danitso/terraform-provider-proxmox/files/4938604/main.tf.txt)
Regarding proxmox itself, the setup is as follows:
- The proxmox host has an lvm-thinpool storage volume group named "datastore1"
- The VM template with ID 900 has a single raw SCSI disk attached, stored on "datastore1"

I believe the following happens:
When cloning from the template, both cloned disks are also created on datastore1. The subsequent move operation tries to move the cloned disks from datastore1 to datastore1, which yields the error message described above.

**Comment by luhahn** [comment]
@dlandtwing Yes, the api supports a target storage. But if you, for example, want to clone from another node, the template needs to be on a shared datastore (For this is NFS). When i want to clone i only can select the nfs store on the target node. 

The move part enables us to move the disk to the local storage. I will have a look into the error. Maybe i just need to add some checks to ignore the move part in special cases :) 

Thanks for the testing!

**Comment by luhahn** [comment]
@dlandtwing Okay, so I added two checks to prevent this behavior. First you can now set datastore_id = "" to prevent the move disk part. Second the error "move to same datastore" should now be ignored and the disk is considered to be moved correctly in the provider. 

As i am currently on vacation and do not have access to my testcluster, can you test both options for me?

File: proxmox/virtual_environment_vm.go [hunk 1]
```diff
@@ -130,29 +157,188 @@ func (c *VirtualEnvironmentClient) GetVMStatus(nodeName string, vmID int) (*Virt
 	return resBody.Data, nil
 }
 
+// MoveVMDisk moves a virtual machine disk.
+func (c *VirtualEnvironmentClient) MoveVMDisk(nodeName string, vmID int, d *VirtualEnvironmentVMMoveDiskRequestBody) error {
+	taskID, err := c.MoveVMDiskAsync(nodeName, vmID, d)
```
Comment (dlandtwing) [inline proxmox/virtual_environment_vm.go:162]: HTTP 500 error "you can't move to the same storage with same format" is returned here

File: proxmox/virtual_environment_vm.go [hunk 2]
```diff
@@ -130,29 +157,188 @@ func (c *VirtualEnvironmentClient) GetVMStatus(nodeName string, vmID int) (*Virt
...
+
+	if err != nil {
+		return err
+	}
+
+	err = c.WaitForNodeTask(nodeName, *taskID, 86400, 5)
+
+	if err != nil && !strings.Contains(err.Error(), "you can't move to the same storage with same format") {
```
Comment (dlandtwing) [inline proxmox/virtual_environment_vm.go:170]: error handling code is not executed, because method has already returned

**Comment by dlandtwing** [comment]
@luhahn thanks a lot for the explanation, that makes sense.

I've tested setting `datastore = ""` in the disk block, works like a charm, thank you!

When setting datastore to the same datastore that the cloned disk is in, the `Received an HTTP 500 response - Reason: you can't move to the same storage with same format` is not caught. I'll add a review to explain what I think happens.
Happy vacation!

File: proxmox/virtual_environment_vm.go (hunk 2 above)
Comment (luhahn) [reply proxmox/virtual_environment_vm.go]: True, i didn't noticed that. Should be fixed now :) Thanks for testing

**Comment by luhahn** [comment]
@dlandtwing i would be glad if you can check if it works correctly now :)

**Comment by luhahn** [comment]
@danitso-dp haven't heard of you are you fine :)?

**Comment by danitso-dp** [comment]
@luhahn I'm still around. I'll get back to this project soon :)

**Comment by luhahn** [comment]
Found out, that multi disk support was not working for clones. Fixed that in another branch. Since this PR is already quite big, shall we open another one when this one is merged? Or shall I include it here. 
It also treats disks now entirely as maps so we can't mess up the indices for disks

**Comment by luhahn** [comment]
if anyone is interested, here's the current state of this branch for amd64 Including multi disk support:
https://novumrgi.github.io/providers/proxmox/1.0.1/linux_amd64/terraform-provider-proxmox_v1.0.1

**Comment by jescarri** [comment]
hey @luhahn I tried your provided build and I get this error:

````
* proxmox_virtual_environment_vm.workers.1: Disk resize fails requests size (50G) is lower than current size (2252M)
* proxmox_virtual_environment_vm.masters[1]: 1 error(s) occurred:
````
tf debug:

````
2020-10-14T21:00:40.120Z [DEBUG] plugin.terraform-provider-proxmox: 2020/10/14 21:00:40 [DEBUG] Values: map[arch:[x86_64] balloon:[0] cipassword:[changeme] ciuser:[ubuntu] cores:[1] cpu:[cputype=qemu64,flags=+aes] cpuunits:[2048] delete:[net3,net4,net5,net6,net7] ide2:[file=local-lvm:cloudinit,media=cdrom] ipconfig0:[gw=192.168.2.1,ip=192.168.2.105/24] ipconfig1:[ip=10.10.3.105/24] ipconfig2:[ip=10.10.4.105/24] memory:[8000] nameserver:[1.1.1.1] net0:[model=virtio,bridge=vmbr0] net1:[model=virtio,bridge=vmbr1] net2:[model=virtio,bridge=vmbr2] onboot:[0] searchdomain:[identitylabs.mx] sockets:[4] sshkeys:[ssh-rsa%20AAAAB3NzaC1yc2EAAAADAQABAAACAQDQqkXbtzu0cTJT2SKQz%2BABwQT2YKUNpVaqLPKrGfV%2BUAaGkRGdb7k8zMY7iDd6gn9RDm5hLRykWUmZnnLMRoBv2AKoryBwJe363kXGE59Ytc2m35tv7mHx2pv54MVOQcOj11aVmm0pV%2Fjq8QKiF4CVgTg6ZM5Nr2sTxoWEW8ucL8PpD%2B11FKuK1ul09XOpsODSPs%2F8ruZXUT8mSIF98%2BE6Rc88%2B8n%2B3zmC7Gv0EHDRQh3OXfmb63k3p0aWsfQeWUWwA2iTkHisJo05Zn%2BEhrzcoNtEY4mf3YmbQlZQXJauRljKU6n64WauK2u1OW7Oi%2FHqF6wEdUh4qlZFh8Mm2HSNptuYeCgecixlZ9Ry4gZYW9WZelfg0hePSyhGy%2BP%2B%2BU1B6pydsNLx%2BHNu8DBCNId0bbiW%2FF9ebz0E3AzIp8hkeqSgESIaMuw7HJ%2Bi6i7b4v6i6nPWYUMzg0bfP3atdkhLhrkIb05oS2GKeIJ5qTUoRCOjjLIL4Uhm%2FaHDXugMMXJ7D%2F4oyyoFhAxKLVdK3melhnSzoAfYMDu09u1d90LBK3rbZUCwNlsIeLFb7hbB7xnitMbWH9hBQalax94fR9PSXVct117oNoSeZucegcxo18f2%2FHGL8uS208FJg4tsGFzqYaRIqJzevO450fh92IGpccWM%2BTU37IWemT0MP5ekvw%3D%3D%20cardno%3A000606460988]]
2020-10-14T21:00:40.121Z [DEBUG] plugin.terraform-provider-proxmox: 2020/10/14 21:00:40 [DEBUG] Added request body to HTTP PUT request (path: nodes/pve01/qemu/5000/config) - Body: arch=x86_64&balloon=0&cipassword=changeme&ciuser=ubuntu&cores=1&cpu=cputype%3Dqemu64%2Cflags%3D%2Baes&cpuunits=2048&delete=net3%2Cnet4%2Cnet5%2Cnet6%2Cnet7&ide2=file%3Dlocal-lvm%3Acloudinit%2Cmedia%3Dcdrom&ipconfig0=gw%3D192.168.2.1%2Cip%3D192.168.2.105%2F24&ipconfig1=ip%3D10.10.3.105%2F24&ipconfig2=ip%3D10.10.4.105%2F24&memory=8000&nameserver=1.1.1.1&net0=model%3Dvirtio%2Cbridge%3Dvmbr0&net1=model%3Dvirtio%2Cbridge%3Dvmbr1&net2=model%3Dvirtio%2Cbridge%3Dvmbr2&onboot=0&searchdomain=identitylabs.mx&sockets=4&sshkeys=ssh-rsa%2520AAAAB3NzaC1yc2EAAAADAQABAAACAQDQqkXbtzu0cTJT2SKQz%252BABwQT2YKUNpVaqLPKrGfV%252BUAaGkRGdb7k8zMY7iDd6gn9RDm5hLRykWUmZnnLMRoBv2AKoryBwJe363kXGE59Ytc2m35tv7mHx2pv54MVOQcOj11aVmm0pV%252Fjq8QKiF4CVgTg6ZM5Nr2sTxoWEW8ucL8PpD%252B11FKuK1ul09XOpsODSPs%252F8ruZXUT8mSIF98%252BE6Rc88%252B8n%252B3zmC7Gv0EHDRQh3OXfmb63k3p0aWsfQeWUWwA2iTkHisJo05Zn%252BEhrzcoNtEY4mf3YmbQlZQXJauRljKU6n64WauK2u1OW7Oi%252FHqF6wEdUh4qlZFh8Mm2HSNptuYeCgecixlZ9Ry4gZYW9WZelfg0hePSyhGy%252BP%252B%252BU1B6pydsNLx%252BHNu8DBCNId0bbiW%252FF9ebz0E3AzIp8hkeqSgESIaMuw7HJ%252Bi6i7b4v6i6nPWYUMzg0bfP3atdkhLhrkIb05oS2GKeIJ5qTUoRCOjjLIL4Uhm%252FaHDXugMMXJ7D%252F4oyyoFhAxKLVdK3melhnSzoAfYMDu09u1d90LBK3rbZUCwNlsIeLFb7hbB7xnitMbWH9hBQalax94fR9PSXVct117oNoSeZucegcxo18f2%252FHGL8uS208FJg4tsGFzqYaRIqJzevO450fh92IGpccWM%252BTU37IWemT0MP5ekvw%253D%253D%2520cardno%253A000606460988
2020-10-14T21:00:40.814Z [DEBUG] plugin.terraform-provider-proxmox: 2020/10/14 21:00:40 [DEBUG] WARNING: Unhandled HTTP response body: {"data":null}
2020-10-14T21:00:40.814Z [DEBUG] plugin.terraform-provider-proxmox: 2020/10/14 21:00:40 [DEBUG] Performing HTTP GET request (path: nodes/pve01/qemu/5000/config)
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalWriteState
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalApplyProvisioners
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalIf
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalWriteState
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalWriteDiff
2020/10/14 21:00:40 [TRACE] root: eval: *terraform.EvalApplyPost
2020/10/14 21:00:40 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* proxmox_virtual_environment_vm.masters: Disk resize fails requests size (50G) is lower than current size (2252M)
2020/10/14 21:00:40 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* proxmox_virtual_environment_vm.masters: Disk resize fails requests size (50G) is lower than current size (2252M)
2020/10/14 21:00:40 [TRACE] [walkApply] Exiting eval tree: proxmox_virtual_environment_vm.masters
2020/10/14 21:00:40 [TRACE] dag/walk: upstream errored, not walking "meta.count-boundary (count boundary fixup)"
2020/10/14 21:00:40 [TRACE] dag/walk: upstream errored, not walking "provider.proxmox (close)"
2020/10/14 21:00:40 [TRACE] dag/walk: upstream errored, not walking "root"
2020/10/14 21:00:40 [TRACE] Preserving existing state lineage "206ba916-48f0-9cf4-df1f-0c286611077f"

Error: Error applying plan:

1 error(s) occurred:

* proxmox_virtual_environment_vm.masters: 1 error(s) occurred:

* proxmox_virtual_environment_vm.masters: Disk resize fails requests size (50G) is lower than current size (2252M)

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
````

tf code:

````
resource "proxmox_virtual_environment_vm" "masters" {
  name      = "cp0${count.index + 1}.potato"
  node_name = "pve01"
  count     = 1
  vm_id     = "${5000 + count.index}"

  clone {
    vm_id   = 9001
    retries = 10
  }

  memory {
    dedicated = 8000
  }

  disk {
    datastore_id = "local-lvm"
    size         = 50

    interface = "scsi0"
  }

  cpu {
    cores   = 1
    sockets = 4
    units   = 2048
    flags   = ["+aes"]
  }

  network_device {
    bridge = "vmbr0"
    model  = "virtio"
  }

  network_device {
    bridge = "vmbr1"
    model  = "virtio"
  }

  network_device {
    bridge = "vmbr2"
    model  = "virtio"
  }

  initialization {
    #datastore_id = "local-lvm"

    user_account {
      keys     = ["${trimspace(var.ssh_key)}"]
      password = "changeme"
      username = "ubuntu"
    }

    dns = {
      domain = "identitylabs.mx"
      server = "1.1.1.1"
    }

    ip_config {
      ipv4 {
        address = "${format("%s/24",element(var.master-ips,count.index))}"
        gateway = "192.168.2.1"
      }
    }

    ip_config {
      ipv4 {
        address = "${format("%s/24",element(var.master-nfs-ips,count.index))}"
      }
    }

    ip_config {
      ipv4 {
        address = "${format("%s/24",element(var.master-iscsi-ips,count.index))}"
      }
    }
  }

  lifecycle {
    ignore_changes = ["disk", "initialization"]
  }
}
`````

**Comment by luhahn** [comment]
@jescarri seems to be an error in formatting MB vs GB. I'll have a look at it, thanks for finding this :)

**Comment by danitso-dp** [comment]
@luhahn any progress on the latest changes?

**Comment by luhahn** [comment]
havent had time to look yet, will check today if i find something

**Comment by luhahn** [comment]
@danitso-dp ok i found the error. However I've another branch which includes the multi disk support. Should I integrate the branch into this PR? Or do you want another seperate PR for this?

**Comment by luhahn** [comment]
@danitso-dp btw it would be great if we can merge this soon, because rebasing and resolving conflicts is getting annoying, since this PR is huge :D

**Comment by luhahn** [comment]
For more information about the multidisk support have a look here:
https://github.com/luhahn/terraform-provider-proxmox/tree/multidisk_support
It is based on this PR, so i did not create a new one yet.

**Comment by luhahn** [comment]
@danitso-dp Any news when we can merge this?

**Comment by ryan4yin** [comment]
we need this feature

**Comment by danitso-dp** [comment]
@luhahn sorry for having been away this long. I haven't had time to check up on the recent changes but would it make sense to provide a new clean PR with both this feature and the multi disk feature? I assume you're already using both on your end

**Comment by danitso-dp** [comment]
@luhahn I'll be working on merging all the changes between Christmas and New Year. I'll collect it all in a new PR for everyone to test before merging and creating a new release. The PR for getting it ready for Terraform Registry will be included as well.

**Comment by danitso-dp** [comment]
Sorry for the delay, which was sadly caused by some unexpected work. However, I'll continue my work on the PR later today and get back to you.

**Comment by luhahn** [comment]
Ah yes, i can provide a new PR with both features.

**Comment by luhahn** [comment]
@danitso-dp please see https://github.com/danitso/terraform-provider-proxmox/pull/60

I've squashed a few commits into one. And yes this version is already running in our clusters. 

This PR can be closed now without merging :)

**Comment by danitso-dp** [comment]
@luhahn great work. I'll continue to get the new version ready with the new PR included.
